contracts/data/BinaryHeap.sol

```diff
@@ -0,0 +1,320 @@
+        return heap._values[index];
+    }
+
+    function _contains(Heap storage heap, bytes32 value)
+        private
+        view
+        returns (bool)
+    {
+        return heap._indexes[value] != 0;
+    }
+
+    function _indexOf(Heap storage heap, bytes32 value)
+        private
+        view
+        returns (uint256)
+    {
+        unchecked {
+            return heap._indexes[value] - 1;
+        }
+    }
+
+    function _length(Heap storage heap) private view returns (uint256) {
+        return heap._values.length;
+    }
+
+    function _root(Heap storage heap) private view returns (bytes32) {
+        return _at(heap, 0);
+    }
+
+    function _add(Heap storage heap, bytes32 value) private returns (bool) {
+        if (!_contains(heap, value)) {
+            heap._values.push(value);
+            heap._indexes[value] = _length(heap);
+            _heapify(heap);
+
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    function _remove(Heap storage heap, bytes32 value) private returns (bool) {
+        if (_contains(heap, value)) {
+            uint256 index = _indexOf(heap, value);
+
+            // move node with last element in the tree, then remove it
+            _swap(heap, index, _length(heap) - 1);
+
+            heap._values.pop();
+            delete heap._indexes[value];
+
+            _heapify(heap);
+
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    function _heapify(Heap storage heap) private {
+        uint256 len = _length(heap);
+        if (len > 1) {
+            unchecked {
+                uint256 index = ((len) / 2) - 1;
+                while (index >= 0) {
+                    _maxHeapify(heap, len, index);
+                    if (index == 0) break;
+                    --index;
+                }
+            }
+        }
+    }
+
+    function _maxHeapify(
+        Heap storage heap,
+        uint256 len,
+        uint256 index
+    ) private {
+        uint256 largest = index;
+
+        unchecked {
+            uint256 left = 2 * index + 1;
+            uint256 right = 2 * index + 2;
+
+            if (left < len && heap._values[largest] < heap._values[left]) {
+                largest = left;
+            }
+
+            if (right < len && heap._values[largest] < heap._values[right]) {
+                largest = right;
+            }
+        }
+
+        if (largest != index) {
+            // swap until the largest node is the root node
+            _swap(heap, index, largest);
+            _maxHeapify(heap, len, largest);
+        }
+    }
+
+    function _swap(
+        Heap storage heap,
+        uint256 a,
+        uint256 b
+    ) private {
+        (heap._indexes[heap._values[a]], heap._indexes[heap._values[b]]) = (
+            heap._indexes[heap._values[b]],
+            heap._indexes[heap._values[a]]
+        );
+
+        (heap._values[a], heap._values[b]) = (heap._values[b], heap._values[a]);
```

Comment:
I believe this could be made more efficient by storing a local reference to `_indexes` and `_values`, rather than accessing via `heap` each time.

Reply:
Same might be true for `_values` in the `_maxHeapify` function above.

Reply:
> I believe this could be made more efficient by storing a local reference to `_indexes` and `_values`, rather than accessing via `heap` each time.

this showed substantial gas savings https://github.com/solidstate-network/solidstate-solidity/pull/134/commits/df978f245df99fa3073d68abfb9aabb4e8dc1d89

Reply:
> Same might be true for `_values` in the `_maxHeapify` function above.

https://github.com/solidstate-network/solidstate-solidity/pull/134/commits/b9e9dc14ed0f76a2aaae6d3e23e7c84bcb941415